substrate/client/network/sync/src/strategy/warp.rs

```diff
@@ -464,12 +474,30 @@ where
 			targets.sort();
 			let median = targets[targets.len() / 2];
 			let threshold = std::cmp::max(median, min_best_number.unwrap_or(Zero::zero()));
-			// Find a random peer that is synced as much as peer majority and is above
+			// Find a random available peer that is synced as much as peer majority and is above
 			// `min_best_number`.
-			for (peer_id, peer) in self.peers.iter_mut() {
-				if peer.state.is_available() && peer.best_number >= threshold {
-					peer.state = new_state;
-					return Some(*peer_id)
+			for peer_id in self.peer_pool.available_peers() {
+				if let Some(peer) = self.peers.get_mut(&peer_id) {
+					if peer.state.is_available() && peer.best_number >= threshold {
+						if self.peer_pool.try_reserve_peer(&peer_id) {
```

Comment:
I think this iterating over available peers/reserving a peer has to be refactored. The availability state of a peer should be stored in one place (`PeerPool`) and `try_reserve_peer()` should just be `reserved_peer()` and made infallible. I was playing around with some code that would allow doing that with an iterator but it went nowhere because of the `MutexGuard` we hold.

I think there are two ways to fix this. `PeerPool` has an explicit `lock()` function and each strategy locks it when it wants to query next available peer. This way you can iterate over `PeerPool`s available peers, check whether the next peer satisfies the best number check and if it does, `reserved_peer()` *always* succeeds in reserving the peer since we hold a lock.

Another way, which I think might be worth exploring, is storing some additional state to `PeerPool` which is utilized by all strategies (such as the best number) and then having a function `reserve_next_available_peer(target: B::Number) -> Option<PeerId>` which iterates over all available peers and tries to find a peer whose `best_number > target`, reserves it and returns its `PeerId`. Whether the idea of storing additional information to `PeerPool` is too leaky remains to be seen.

Making `try_reserve_peer()` infallible would make much of the related code cleaner and technically (with Sync 2.0 in mind) failing to reserve a peer which was just previously iterated over as an available peer is not worth a `debug_assert!()` because Sync 1.0 and 2.0 might iterate over the available peers at the same time and try to reserve the same peer, only one succeeding so there should be a way to retry and not crash.

Reply:
I followed the first approach, as different strategies use different approaches for peer selection. For example, `ChainSync` selects peers in a completely different way than warp & state sync.